### ISSUE TITLE
Password with special characters (€) not  working ( 401 from server )

### ISSUE THREAD

**Issue by ubabap** [abapGit/abapGit#4800]
Hello together,

for abapGit v1.108 (both standalone and dev, and V1.107 standalone got the issue . Different SAP Systems tried ), SAPGUI 760 , SAP_BASIS 750 sp5, 

I changed my password to one containing an €, and always did get a 401 error when using abapGit to connect to a remote Git repository. 
I can't see any issue with the password till it is passed to the remote Git.
Remote Git is a bitbucket (server version, no cloud ) . 
No issues with login there on Web client, 
or local git Client (Sourcetree).

All the time app password did work from abapGit.
(No exits are used, Basisc Authentication is used  )

It does not to be a Popup issue  -> [4055] (https://github.com/abapGit/abapGit/issues/4055). 

Workaorund: Changing back to a password using other special character solved the issue.
(I also asked a colleague to check, and there was the same issue. I don't know the used characters and he said there would be no change since yesterday, and the error would be new...  Maybe, there is something else, but i can't reproduce that )

If anybody knows, which letters are allowed/not allowed , maybe a check with a suitable message could be suggested .  
( i can't imagine it is an open issue to be solved generally so maybe somehow add checks via exit locally . Maybe there is something already thought about special local password checks or any local adaption for a future popup  [#783](https://github.com/abapGit/abapGit/issues/783)   . Or there is already  a mechanism? The repsonse 401 is not correct .I need to   reproduce the situation first. I still submit it for now, just in case someone else encounters the same. )

**Comment by mbtools** [comment]
In general, abapgit uses basic authentication for HTTP with standard classes. Put a break-point at `cl_http_request->IF_HTTP_REQUEST~SET_AUTHORIZATION` and check if password still looks ok. If yes, it's certainly not an abapGit issue and you should check Kernel patches. 

If it doesn't stop, the server requested digest access, where the user/password has to be hashed with MD5. We would have to check `zcl_abapgit_http_digest`. 

Is your system Unicode or not?

**Comment by ubabap** [comment]
The systems tested are all  Unicode. 
I did look at the password in the code till after any point in the abapGit code. It is  was looking ok. I did not  follow it into SAP standard Code. 
I am not able to request a kernel patch (takes ages, till our Basis does so and i doubt, that this is the root cause. I will not do so, wihtout having a SAP note or Notficication for that) .
Unfortunately i am not able to switch to antoher password. I think, there might be a unknwon corporate policy that i can't change it for a while. It took me more than a fullday, till i could any password that was accepted to get rid of the issue. I
 will try to reproduce it at Monday and check again the concrete contents .

**Comment by mbtools** [comment]
I was able to reproduce this problem using AG and Github. It does not work when using special characters in the password.

It's not an SAP-specific problem since there's no clear definition as to how username and password must be encoded when using Basic Authentication RFC 2617 (https://stackoverflow.com/questions/7242316/what-encoding-should-i-use-for-http-basic-authentication).

SAP has documented this limitation in SAP Notes [1240796](https://launchpad.support.sap.com/#/notes/1240796) (and mentioned in SAP Note [2910375](https://launchpad.support.sap.com/#/notes/2910375) as well). 

Since 2015, there's RFC 7617 which clarifies that UTF-8 shall be used. However, at least up to Kernel 753 it seems SAP has not implemented this spec. I will check with newer Kernels later.

In summary, you can't use special characters in your password. 

Alternatives: Use tokens (works with Github).

**Comment by mbtools** [comment]
Unfortunately, same result with SAP Basis 7.54 and Kernel 7.77 patch 200. 

To clarify for everyone, "special characters" in this context refers to non-ASCII characters. ASCII characters - including extended ASCII should be working well in passwords (hex 20 to FF, https://en.wikipedia.org/wiki/ISO/IEC_8859-1). 

The Euro symbol however is not included since it's Unicode U+20AC.

**Comment by mbtools** [comment]
Todo: Update https://docs.abapgit.org/guide-authentication.html

**Comment by ubabap** [comment]
Thank's alot for the reproducing and updating the help. I think, that Information is perfect to close the Issue.
 Using App Id in bitbucket, a  replacement for the users password , did work all the time as a workaround.  
 (Other local Clients than a  SAPSystem  do not have that issue, also with Basic Authentification. )

**Comment by mbtools** [comment]
do you have an example of where it works?

**Comment by ubabap** [comment]
I just tried again using Sourcetree and a password containing non-Ascii.
That is working in Sourcetree, Git Bash connection fails .

 According to settings in sourcetree Basic Authentication is used for bitbucket account, Rest and Git connection ( I do not know, when it is using them . Just listing the options available for stored passwords )

**Comment by ubabap** [comment]
Correction: Also GitBash does work with Non Ascii password . At least for a new connection.

**Comment by mbtools** [comment]
thanks. yes, other clients have their own implementation or library for authentication. until SAP fixes this limitation, unfortunately it won't work with abapGit.
